### PR TITLE
Implementing OmitEmpty flag to skip zero-value fields.

### LINKED ISSUE #22: Omit empty as an option
Often I'm working with large structs which have fields which are usually empty, but which are of a lot of interest if not empty.  I would love there to be an option to omit empty fields.  The problem of course is how we define empty.

I would suggest that simply considering the zero or nil state of any field to be an empty field.  So, "" would be empty for a string, and 0 would be empty for an int.

### PR DESCRIPTION
**Pull request by peterebden**
Found that I had a need for this too, so thought I'd have a crack at it.

Resolves #22

### CODE REVIEW

**Comment by peterebden** [comment]
Hmmmm, the Travis failure is not related to this PR, go vet is complaining about unsafe.Pointer in a couple of other places. Let me know if there's anything I should do for it...

File: spew/dump_test.go [hunk 1]
```diff
@@ -1040,3 +1040,46 @@ func TestDumpSortedKeys(t *testing.T) {
...
+	}
+	s1 := s{S1: &s{S2: &s{Int: 5}}}
+	actual := cfg.Sdump(s1)
+	expected := "(spew_test.s) {\n" +
+		"S1: (*spew_test.s)({\n" +
+		"S2: (*spew_test.s)({\n" +
+		"Int: (int) 5,\n" +
+		"}),\n" +
+		"}),\n" +
+		"}\n"
```
Comment (shurcooL) [inline spew/dump_test.go]: Have you considered using a [raw string](https://golang.org/ref/spec#raw_string_lit) here? I think it’d be much more readable.

Comment (peterebden) [reply spew/dump_test.go]: Makes sense, I was trying to match the style of the other tests, but you're right here I think.

**Comment by mediocregopher** [comment]
Any update on this? I'd really love this feature :3

File: spew/config.go [hunk 2]
```diff
@@ -98,6 +98,10 @@ type ConfigState struct {
 	// be spewed to strings and sorted by those strings.  This is only
 	// considered if SortKeys is true.
 	SpewKeys bool
+
+	// OmitEmpty specifies that empty struct fields should be omitted.
+	// Empty fields are any that have the zero value for their type.
+	OmitEmpty bool
```
Comment (maxatome) [inline spew/config.go:104]: I prefer `OmitZeroStructFields` as only struct fields are impacted and I propose to not consider the same way an empty slice/map vs a `nil` one. IMO only `nil` ones should be omitted, just to allow to see the difference between a `nil` (a.k.a non allocated) and an empty item (allocated but empty).
Following the same logic, empty array should not be omitted, so the behavior is consistent with non `nil` empty slice/map as I described above. Moreover, empty arrays are rare, so no need to try to hide them.
`nil` functions can be omitted too.

File: spew/dump.go [hunk 3]
```diff
@@ -507,3 +511,26 @@ get the formatted result as a string.
...
+		return v.Len() == 0
+	case reflect.Bool:
+		return !v.Bool()
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return v.Int() == 0
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return v.Uint() == 0
+	case reflect.Float32, reflect.Float64:
+		return v.Float() == 0
+	case reflect.Interface, reflect.Ptr, reflect.Chan:
```
Comment (maxatome) [inline spew/dump.go:528]:
```suggestion
	case reflect.Map, reflect.Slice, reflect.Interface, reflect.Ptr, reflect.Chan, reflect.Func:
```

File: spew/dump.go [hunk 4]
```diff
@@ -507,3 +511,26 @@ get the formatted result as a string.
 func Dump(a ...interface{}) {
 	fdump(&Config, os.Stdout, a...)
 }
+
+// isZero is a helper function that checks whether a reflect.Value has the zero value for its type.
+func isZero(v reflect.Value) bool {
+	switch v.Kind() {
+	case reflect.Array, reflect.Map, reflect.Slice, reflect.String:
```
Comment (maxatome) [inline spew/dump.go:518]:
```suggestion
	case reflect.String:
```
and never ignore empty arrays as explained above.

File: spew/internal_test.go [hunk 5]
```diff
@@ -82,3 +82,13 @@ func TestInvalidReflectValue(t *testing.T) {
...
+
+func TestIsZero(t *testing.T) {
+	if !isZero(reflect.ValueOf(0)) {
+		t.Errorf("isZero should be true for 0")
+	}
+	var v reflect.Value
+	if isZero(v) {
+		t.Errorf("isZero should be false for an invalid Value")
+	}
+}
```
Comment (maxatome) [inline spew/internal_test.go:94]:
```suggestion
func TestIsZero(t *testing.T) {
	// isZero true
	for _, v := range []interface{}{
		false,
		int8(0), int16(0), int32(0), int64(0), 0,
		uint8(0), uint16(0), uint32(0), uint64(0), uint(0),
		float32(0), float64(0),
		complex64(0), complex128(0),
		"",
		([]int)(nil), (map[int]int)(nil), (func())(nil), (chan int)(nil), (*int)(nil),
		nil,
	} {
		if !isZero(reflect.ValueOf(v)) {
			t.Errorf("isZero should be true for %[1]v (%[1]T)", v)
		}
	}

	// isZero false
	for _, v := range []interface{}{
		true,
		int8(1), int16(1), int32(1), int64(1), 1,
		uint8(1), uint16(1), uint32(1), uint64(1), uint(1),
		float32(1), float64(1),
		complex64(1), complex128(1),
		"!",
		[]int{}, map[int]int{}, func() {}, make(chan int), &[]int{},
	} {
		if isZero(reflect.ValueOf(v)) {
			t.Errorf("isZero should be false for %[1]v (%[1]T)", v)
		}
	}
}
```

**Comment by peterebden** [comment]
To be honest I'd forgotten about this PR.

I can see the motivation for your proposed changes, but I'm not sure if printing empty slices would have a negative impact on what I wanted it for originally. I guess I'm kind of agnostic on it, but this is very stale now; is this project still interested in new contributions? If not there's no point discussing further really...

**Comment by maxatome** [comment]
@davecgh @dajohi do you accept new contributions?

**Comment by awkaiser** [comment]
I agree with @maxatome's suggested changes 👍

**Comment by khpeek** [comment]
@peterebden This change would still be most welcome in my view! Some structs I would like to debug have a plethora of fields which are unset and irrelevant to what I'm debugging; essentially I would like to print them out as if they were in JSON format with the `json:",omitempty"` struct tag applied everywhere, which as I understand it this contribution achieves.

**Comment by peterebden** [comment]
I had forgotten that this was still a thing to be honest.

I don't have any particular need for this any more (it was relevant to me in 2018, not really any more). Happy if someone else wants to copy this & continue with it, although as I mentioned above it seems this project isn't looking for contributions any more. Good luck :)

**Pull request closed without merging**